File: app/src/main/java/com/nytimes/android/external/register/Signer.java
package com.nytimes.android.external.register;

import android.support.annotation.Nullable;

import com.google.common.base.Optional;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.PrivateKey;
import java.security.Signature;
import java.security.SignatureException;

import javax.inject.Inject;

public class Signer {

    @Inject
    @Nullable
    Optional<PrivateKey> privateKey;

    @Inject
    @Nullable
    Optional<Signature> signature;

    public Signer(@Nullable Optional<PrivateKey> privateKey, @Nullable Optional<Signature> signature) {
        this.privateKey = privateKey;
        this.signature = signature;
    }

    public String signData(String unsignedData) throws InvalidKeyException, SignatureException {
        String signedData = "";
        if (signature.isPresent()) {
            signature.get().initSign(privateKey.orNull());
            signature.get().update(unsignedData.getBytes(StandardCharsets.UTF_8));
            signedData = new String(signature.get().sign(), StandardCharsets.UTF_8);
        }
        return signedData;
    }
}

File: app/src/main/java/com/nytimes/android/external/register/Signer.kt
package com.nytimes.android.external.register

import java.nio.charset.StandardCharsets
import java.security.InvalidKeyException
import java.security.PrivateKey
import java.security.Signature
import java.security.SignatureException

class Signer(private val privateKey: PrivateKey?,
             private val signature: Signature?) {

    @Throws(InvalidKeyException::class, SignatureException::class)
    fun signData(unsignedData: String): String {
        return signature?.let {
            it.initSign(privateKey)
            it.update(unsignedData.toByteArray(StandardCharsets.UTF_8))
            String(it.sign(), StandardCharsets.UTF_8)
        } ?: ""
    }
}
